samcli/lib/init/template_modifiers/application_insights_template_modifier.py

```diff
@@ -0,0 +1,97 @@
+"""
+Class used to parse and update template when application-insights is enabled
+"""
+import logging
+from typing import Any
+from samcli.lib.init.template_modifiers.cli_template_modifier import TemplateModifier
+from samcli.lib.utils.resources import (
+    AWS_APPLICATION_INSIGHTS,
+    AWS_RESOURCE_GROUP
+)
+
+LOG = logging.getLogger(__name__)
+
+
+class ApplicationInsightsTemplateModifier(TemplateModifier):
+    import ruamel.yaml
+    from ruamel.yaml import YAML
+    from ruamel.yaml.comments import CommentedMap
+
+    TYPE_KEY = "Type"
+    RESOURCES_KEY = "Resources"
+    PROPERTIES_KEY = "Properties"
+    NAME_KEY = "Name"
+    RESOURCE_QUERY_KEY = "ResourceQuery"
+    RESOURCE_GROUP_NAME_KEY = "ResourceGroupName"
+    AUTO_CONFIG_ENABLED_KEY = "AutoConfigurationEnabled"
+    DEPENDS_ON_KEY = "DependsOn"
+    CFN_STACK_TYPE = "CLOUDFORMATION_STACK_1_0"
+    RESOURCE_GROUP_REF = "ApplicationResourceGroup"
+    APPLICATION_INSIGHTS_REF = "ApplicationInsightsMonitoring"
+    AUTO_CONFIG_VALUE = "true"
+    RESOURCE_GROUP_NAME = {"Fn::Join": ["", ["ApplicationInsights-SAM-", {"Ref": "AWS::StackName"}]]}
+
+    # set ignore aliases to true. This configuration avoids usage yaml aliases which is not parsed by CloudFormation.
+    class NonAliasingRTRepresenter(ruamel.yaml.representer.RoundTripRepresenter):
+        def ignore_aliases(self, data):
+            return True
+
+    def __init__(self, location):
+        self.yaml = ApplicationInsightsTemplateModifier.YAML()
+        self.yaml.Representer = ApplicationInsightsTemplateModifier.NonAliasingRTRepresenter
+        super().__init__(location)
+
+    def _get_template(self) -> Any:
+        with open(self.template_location) as file:
+            return self.yaml.load(file)
+
+    def _update_template_fields(self):
+        """
+        Add new resources to SAM template
+        """
+        self._add_app_insights_monitoring_section()
+
+    def _add_app_insights_monitoring_section(self):
+        resourceGroup = {
```

Comment:
should this be versioned? how will this change over time if it does? will there be new cfn properties?

Reply:
I don't think it needs to be versioned

Reply:
As discussed offline, lets continue the conversation on keeping this section upto date. Since we are injecting new resources from client side tooling.